engines/ic3ia.cpp

```diff
@@ -19,278 +19,256 @@
-      conjuncts.push_back(p);
-    } else {
-      assert(val == solver_false_);
-      conjuncts.push_back(solver_->make_term(Not, p));
-    }
-  }
-  return Conjunction(solver_, conjuncts);
+  // basically a No-Op
+  // no restrictions except that interpolants must be supported
+  // instead of checking explicitly, just let the interpolator throw an
+  // exception better than maintaining in two places
 }
 
-void IC3IA::set_labels()
+void IC3IA::initialize()
 {
-  assert(solver_context_ == 0);  // expecting to be at base context level
-  // set semantics of labels
-  Sort boolsort = solver_->make_sort(BOOL);
-  if (!init_label_) {
-    // frame 0 label is identical to init label
-    init_label_ = frame_labels_[0];
+  super::initialize();
+
+  // add all the predicates from init and property to the abstraction
+  // NOTE: abstract is called automatically in IC3Base initialize
+  UnorderedTermSet preds;
+  get_predicates(solver_, ts_->init(), preds, false);
+  get_predicates(solver_, bad_, preds, false);
+  for (auto p : preds) {
+    add_predicate(p);
   }
-  if (!trans_label_) {
-    trans_label_ = solver_->make_symbol("__trans_label", boolsort);
-    // Note: Using iff instead of implies so that we can negate it
-    // for example in relational preimage generalization
-    solver_->assert_formula(
-        solver_->make_term(Implies, trans_label_, abs_ts_.trans()));
-
-    // add all the predicates from init and property
-    UnorderedTermSet preds;
-    get_predicates(solver_, ts_->init(), preds, false);
-    get_predicates(solver_, bad_, preds, false);
-    for (auto p : preds) {
-      add_predicate(p);
+  // more predicates will be added during refinement
+  // these ones are just initial predicates
+
+  // populate cache for existing terms in solver_
+  UnorderedTermMap & cache = to_solver_.get_cache();
+  Term ns;
+  for (auto s : ts_->statevars()) {
+    // common variables are next states, unless used for refinement in IC3IA
+    // then will refer to current state variables after untiming
+    // need to cache both
+    cache[to_interpolator_.transfer_term(s)] = s;
+    ns = ts_->next(s);
+    cache[to_interpolator_.transfer_term(ns)] = ns;
+  }
+
+  // need to add uninterpreted functions as well
+  // first need to find them all
+  // NOTE need to use get_free_symbols NOT get_free_symbolic_consts
+  // because the latter ignores uninterpreted functions
+  UnorderedTermSet free_symbols;
+  get_free_symbols(ts_->init(), free_symbols);
+  get_free_symbols(ts_->trans(), free_symbols);
+  get_free_symbols(bad_, free_symbols);
+
+  for (auto s : free_symbols) {
+    assert(s->is_symbol());
+    if (s->is_symbolic_const()) {
+      // ignore constants
+      continue;
     }
+    cache[to_interpolator_.transfer_term(s)] = s;
   }
-}
 
-void IC3IA::check_ts() const
-{
-  // this is a complete No-Op for IC3IA
-  // no current limitations (among major theories)
+  // TODO fix generalize_predecessor for ic3ia
```

Comment:
it shouldn't be a problem, I think.  or Maybe I am missing something here.

Reply:
So it was failing, e.g. the formula wasn't unsatisfiable. I think it has something to do with the predicate abstraction? I need to look into it more still. Might be why the MathSAT IC3IA has a bail-out option from the generalization. Even the version for non-deterministic systems isn't guaranteed to be unsat I don't think because we're not using model values.